inkscape/textext: init at 1.8.1

raboof opened the pull request:
###### Motivation for this change

Introduce the 'textext' extension for including LaTeX graphics in Inkscape

Test it with `nix-shell -p "inkscape-with-extensions.override { inkscapeExtensions = [ inkscape-extensions.textext ]; }" --run inkscape`

Alternative approach to #100565

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

raboof committed:
inkscape-extensions.textext: init at 1.8.1

Introduce the 'textext' extension for including LaTeX graphics in Inkscape

Test it with `nix-shell -p "inkscape-with-extensions.override {
inkscapeExtensions = [ inkscape-extensions.textext ]; }" --run inkscape`

Co-Authored-By: Jan Tojnar <jtojnar@gmail.com>

r-rmcgibbo commented:
Result of `nixpkgs-review pr 121711` at d3fce869 run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built successfully:</summary>
  <ul>
    <li>inkscape-extensions.textext</li>
  </ul>
</details>
<details>
  <summary>2 suggestions:</summary>
  <ul>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/python-include-tests.md">python-include-tests</a>

Consider adding a `checkPhase` for tests, or if not feasible, `pythonImportsCheck`.

Near pkgs/applications/graphics/inkscape/extensions.nix:104:0:
```
    |
104 |          description = "Re-editable LaTeX graphics for Inkscape";
    | ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-patch-comment.md">missing-patch-comment</a>

Please add a comment on the line above, explaining the purpose of this patch.
Near pkgs/applications/graphics/inkscape/extensions.nix:61:11:
```
   |
61 |           ./textext-placeholders.patch
   |           ^
```
</li>
  </ul>
</details>

roberth commented:
Please rebase

roberth commented on pkgs/applications/graphics/inkscape/extensions/textext.nix:
```diff
@@ -0,0 +1,75 @@
+    mkdir -p $out/share/inkscape/extensions
+    cp -r dist/textext $out/share/inkscape/extensions
+
+    substituteInPlace $out/share/inkscape/extensions/textext/base.py \
+      --replace PDFLATEX_PLACEHOLDER ${texlive-combined}/bin/pdflatex \
+      --replace LUALATEX_PLACEHOLDER ${texlive-combined}/bin/lualatex
```
Is it possible to do perform this replacement on the source in `postPatch` instead?
That would seem more robust, in case `pyc` files are produced, now or in a possible future.

jtojnar replied:
Even nicer would be using `substituteAll` on the patch itself.

roberth reviewed (commented):
This seems fairly straightforward, so I'm inclined to merge it.
Maybe @jtojnar as inkscape maintainer will see something I don't.

jtojnar commented on pkgs/applications/graphics/inkscape/extensions/textext.nix:
```diff
@@ -0,0 +1,75 @@
+}:
+
+let
+  texlive-combined = texlive.combined.scheme-basic;
```
Does the extension support arbitrary latex packages? In that case this should probably be passed as an argument so it can be changed using `override`?

jtojnar commented on pkgs/applications/graphics/inkscape/extensions/textext.nix:
```diff
@@ -0,0 +1,75 @@
+    # GTK is recommended, but I haven't looked into pkgs.pygobject3 yet
+    # https://nixos.org/manual/nixpkgs/stable/#ssec-gnome-hooks-gobject-introspection
+    python3.pkgs.tkinter
+    texlive-combined
```
Propagating this probably will not do anything.

jtojnar commented on pkgs/applications/graphics/inkscape/extensions/textext.nix:
```diff
@@ -0,0 +1,72 @@
+  patches = [
+    # Make sure we can point directly to pdflatex in the extension,
+    # instead of relying on the PATH (which might not have it)
+    (runCommand "textext-paths.patch" {} ''
```
Sorry, meant `substituteAll` function in Nixpkgs:

https://github.com/NixOS/nixpkgs/blob/291b3ff5af268eb7a656bb11c73f2fe535ea2170/pkgs/desktops/gnome/core/gnome-shell/default.nix#L80-L84

(Will need placeholders marked with `@` in the patch file.

raboof replied:
Ahh, TIL :)

jtojnar commented on pkgs/applications/graphics/inkscape/extensions/textext-paths.patch:
```diff
@@ -0,0 +1,13 @@
+@@ -125,6 +125,8 @@ class TexText(inkex.EffectExtension):
+     def __init__(self):
+ 
+         self.config = Settings("config.json")
```
Won’t this already contain the values, as passed to `setup.py`? Or can this be a user config? Would be nice to clarify in a comment.

raboof replied:
I'm slightly on thin ice here, but indeed I'm pretty sure that's how it works. Added a comment.

jtojnar commented:
Since this is multiple files, it would be nice to move them into a subdirectory.

Also, I tried switching to the pygobject but we will need to add bash to Inkscape’s interpreter list.

```patch
diff --git a/pkgs/applications/graphics/inkscape/extensions/textext.nix b/pkgs/applications/graphics/inkscape/extensions/textext.nix
index 29e001bfca8..4da59dba952 100644
--- a/pkgs/applications/graphics/inkscape/extensions/textext.nix
+++ b/pkgs/applications/graphics/inkscape/extensions/textext.nix
@@ -5,9 +5,12 @@
 , pdflatex
 , lualatex
 , python3
+, wrapGAppsHook
+, gobject-introspection
+, gtk3
 }:
 
-python3.pkgs.buildPythonPackage rec {
+python3.pkgs.buildPythonApplication rec {
   pname = "textext";
   version = "1.3.1";
 
@@ -25,17 +28,36 @@ python3.pkgs.buildPythonPackage rec {
       src = ./textext-paths.patch;
       inherit pdflatex lualatex;
     })
+
+    # Since we are wrapping the extension, we need to change the interpreter
+    # from Python to Bash.
+    ./textext/interpreter.patch
+  ];
+
+  nativeBuildInputs = [
+    wrapGAppsHook
+    gobject-introspection
+  ];
+
+  buildInputs = [
+    gtk3
   ];
 
   propagatedBuildInputs = [
     # GTK is recommended, but I haven't looked into pkgs.pygobject3 yet
     # https://nixos.org/manual/nixpkgs/stable/#ssec-gnome-hooks-gobject-introspection
-    python3.pkgs.tkinter
+    python3.pkgs.pygobject3
   ];
 
+  # strictDeps do not play nicely with introspection setup hooks.
+  strictDeps = false;
+
   # textext doesn't have a 'bdist_wheel' target
   dontUseSetuptoolsBuild = true;
 
+  # Avoid wrapping two times by just using Python’s wrapping.
+  dontWrapGApps = true;
+
   buildPhase = ''
     runHook preBuild
 
@@ -62,6 +84,17 @@ python3.pkgs.buildPythonPackage rec {
     runHook postInstall
   '';
 
+  preFixup = ''
+    # Prepare for wrapping
+    chmod +x "$out/share/inkscape/extensions/textext/__main__.py"
+    sed -i '1i#!/usr/bin/env python3' "$out/share/inkscape/extensions/textext/__main__.py"
+    makeWrapperArgs+=("''${gappsWrapperArgs[@]}")
+  '';
+
+  postFixup = ''
+    wrapPythonProgramsIn "$out/share/inkscape/extensions/textext" "$out $pythonPath"
+  '';
+
   meta = with lib; {
     description = "Re-editable LaTeX graphics for Inkscape";
     homepage = "https://textext.github.io/textext/";
diff --git a/pkgs/applications/graphics/inkscape/extensions/textext/interpreter.patch b/pkgs/applications/graphics/inkscape/extensions/textext/interpreter.patch
new file mode 100644
index 00000000000..39dcdb76eff
--- /dev/null
+++ b/pkgs/applications/graphics/inkscape/extensions/textext/interpreter.patch
@@ -0,0 +1,12 @@
+diff --git a/textext/textext.inx b/textext/textext.inx
+index b1c8835..1cdf7b3 100644
+--- a/textext/textext.inx
++++ b/textext/textext.inx
+@@ -8,6 +8,6 @@
+     </effects-menu>
+   </effect>
+   <script>
+-    <command location="inx" interpreter="python">__main__.py</command>
++    <command location="inx" interpreter="bash">__main__.py</command>
+   </script>
+ </inkscape-extension>
```

jtojnar commented:
There is actually a `shell` interpreter so trying that now.

jtojnar commented:
This works great:

```patch
diff --git a/pkgs/applications/graphics/inkscape/extensions/textext.nix b/pkgs/applications/graphics/inkscape/extensions/textext.nix
index 29e001bfca8..0e7c4299a60 100644
--- a/pkgs/applications/graphics/inkscape/extensions/textext.nix
+++ b/pkgs/applications/graphics/inkscape/extensions/textext.nix
@@ -5,9 +5,13 @@
 , pdflatex
 , lualatex
 , python3
+, wrapGAppsHook
+, gobject-introspection
+, gtk3
+, gtksourceview3
 }:
 
-python3.pkgs.buildPythonPackage rec {
+python3.pkgs.buildPythonApplication rec {
   pname = "textext";
   version = "1.3.1";
 
@@ -25,17 +29,37 @@ python3.pkgs.buildPythonPackage rec {
       src = ./textext-paths.patch;
       inherit pdflatex lualatex;
     })
+
+    # Since we are wrapping the extension, we need to change the interpreter
+    # from Python to Bash.
+    ./textext/interpreter.patch
+  ];
+
+  nativeBuildInputs = [
+    wrapGAppsHook
+    gobject-introspection
+  ];
+
+  buildInputs = [
+    gtk3
+    gtksourceview3
   ];
 
   propagatedBuildInputs = [
-    # GTK is recommended, but I haven't looked into pkgs.pygobject3 yet
-    # https://nixos.org/manual/nixpkgs/stable/#ssec-gnome-hooks-gobject-introspection
-    python3.pkgs.tkinter
+    python3.pkgs.pygobject3
+    # TODO: Required by inkex but is not inherited from inkscape when we use custom Python interpreter.
+    python3.pkgs.lxml
   ];
 
+  # strictDeps do not play nicely with introspection setup hooks.
+  strictDeps = false;
+
   # textext doesn't have a 'bdist_wheel' target
   dontUseSetuptoolsBuild = true;
 
+  # Avoid wrapping two times by just using Python’s wrapping.
+  dontWrapGApps = true;
+
   buildPhase = ''
     runHook preBuild
 
@@ -62,6 +86,17 @@ python3.pkgs.buildPythonPackage rec {
     runHook postInstall
   '';
 
+  preFixup = ''
+    # Prepare for wrapping
+    chmod +x "$out/share/inkscape/extensions/textext/__main__.py"
+    sed -i '1i#!/usr/bin/env python3' "$out/share/inkscape/extensions/textext/__main__.py"
+    makeWrapperArgs+=("''${gappsWrapperArgs[@]}")
+  '';
+
+  postFixup = ''
+    wrapPythonProgramsIn "$out/share/inkscape/extensions/textext" "$out $pythonPath"
+  '';
+
   meta = with lib; {
     description = "Re-editable LaTeX graphics for Inkscape";
     homepage = "https://textext.github.io/textext/";
diff --git a/pkgs/applications/graphics/inkscape/extensions/textext/interpreter.patch b/pkgs/applications/graphics/inkscape/extensions/textext/interpreter.patch
new file mode 100644
index 00000000000..64b39b61760
--- /dev/null
+++ b/pkgs/applications/graphics/inkscape/extensions/textext/interpreter.patch
@@ -0,0 +1,12 @@
+diff --git a/textext/textext.inx b/textext/textext.inx
+index b1c8835..1cdf7b3 100644
+--- a/textext/textext.inx
++++ b/textext/textext.inx
+@@ -8,6 +8,6 @@
+     </effects-menu>
+   </effect>
+   <script>
+-    <command location="inx" interpreter="python">__main__.py</command>
++    <command location="inx" interpreter="shell">__main__.py</command>
+   </script>
+ </inkscape-extension>
```

![image](https://user-images.githubusercontent.com/705123/124610447-ceabc880-de70-11eb-91cd-5af2874e9ed4.png)

raboof commented:
Cool! Do you want to make a new PR to supersede this one? (you're also welcome to force-push to this branch, but I'm not sure github will let us)

jtojnar commented:
I can force push.

raboof commented:
I can confirm this works for me as well! :+1: nice

jtojnar commented:
Done.

Another worrying aspect is that both the config and the produced SVGs contain absolute store path of the preamble:

```json
  "preamble": "/nix/store/0g42vvrpxanilivh4292ky90scvs80z7-inkscape-with-extensions-1.1/share/inkscape/extensions/textext/default_packages.tex",
```
```xml
       ns5:preamble="/nix/store/0g42vvrpxanilivh4292ky90scvs80z7-inkscape-with-extensions-1.1/share/inkscape/extensions/textext/default_packages.tex"
```

Maybe we will need some code that resolves it to `default_packages.tex` on export and to the absolute path on import.

jtojnar commented on pkgs/applications/graphics/inkscape/extensions/textext/default.nix:
```diff
@@ -0,0 +1,111 @@
+  src = fetchFromGitHub {
+    owner = "textext";
+    repo = "textext";
+    rev = "1.3.1";
```
This should be `rev = version`.

BTW, 1.4.0 has been released since.

raboof commented:
> Done.

Thanks!

> Another worrying aspect is that both the config and the produced SVGs contain absolute store path of the preamble:
> 
> ```json
>   "preamble": "/nix/store/0g42vvrpxanilivh4292ky90scvs80z7-inkscape-with-extensions-1.1/share/inkscape/extensions/textext/default_packages.tex",
> ```
> 
> ```
>        ns5:preamble="/nix/store/0g42vvrpxanilivh4292ky90scvs80z7-inkscape-with-extensions-1.1/share/inkscape/extensions/textext/default_packages.tex"
> ```
> 
> Maybe we will need some code that resolves it to `default_packages.tex` on export and to the absolute path on import. And possibly upstream that.

Yeah, that can't really be portable... I haven't looked at the code, but did a quick experiment breaking the preamble in a saved file, and that didn't cause problems: I was still able to edit the file (also the tex part).

jtojnar commented:
I peeked at the code a bit and we would need to use basename on the path here when it is the default one:

https://github.com/textext/textext/blob/2f1e11c6671756d4693ad497d9f5b28d0e51e214/textext/base.py#L367

We probably would not even need to re-add it, thanks to:

https://github.com/textext/textext/blob/2f1e11c6671756d4693ad497d9f5b28d0e51e214/textext/base.py#L235-L236

But we can merge this as is and return to this later if you do not feel like addressing that now.

raboof commented:
> We probably would not even need to re-add it, thanks to:
> 
> https://github.com/textext/textext/blob/2f1e11c6671756d4693ad497d9f5b28d0e51e214/textext/base.py#L235-L236

Aah, that explains why it worked indeed ;)

> But we can merge this as is and return to this later if you do not feel like addressing that now.

I think that'd make sense for now.

jtojnar commented:
> Aah, that explains why it worked indeed ;)

I guess it will only break when the preamble in the `config.json` does not exist (after update and garbage collection).

Also forgot to mention this line where the config is being updated:

https://github.com/textext/textext/blob/2f1e11c6671756d4693ad497d9f5b28d0e51e214/textext/base.py#L417-L420

The file existence check might rescue it but only after the next edit and the empty string in config would have to make the program use the preamble from the default config. We should probably test that.

olynch commented:
I've tested this out from your branch, and I get an error that the GUI library is not found.

olynch commented:
I fixed the problem, but I don't know how to send a PR to a PR... Here are my changes anyways: https://github.com/olynch/nixpkgs/tree/inkscape-textext-init-at-1.3.1

jtojnar commented:
I am unable to reproduce that issue. And your patch should not really fix the issue – if the check is unable to find the GUI library, how could the extension?

jtojnar commented:
Just tried rebasing this and the issue with hardcoded paths happens even without opening any files – if the path in settings (`~/.config/inkscape/extensions/textext/config.json`) is GC’d, the extension will complain:

```
Command /nix/store/cyw4szxxaibdd3jayjf5r18cagjm4kb2-texlive-combined-basic-2021.20210408/bin/lualatex tmp.tex -interaction=nonstopmode -halt-on-error failed: [Errno 2] No such file or directory: '/nix/store/cyw4szxxaibdd3jayjf5r18cagjm4kb2-texlive-combined-basic-2021.20210408/bin/lualatex'
```

raboof commented:
> Just tried rebasing this and the issue with hardcoded paths happens even without opening any files – if the path in settings (`~/.config/inkscape/extensions/textext/config.json`) is GC’d, the extension will complain:
> 
> ```
> Command /nix/store/cyw4szxxaibdd3jayjf5r18cagjm4kb2-texlive-combined-basic-2021.20210408/bin/lualatex tmp.tex -interaction=nonstopmode -halt-on-error failed: [Errno 2] No such file or directory: '/nix/store/cyw4szxxaibdd3jayjf5r18cagjm4kb2-texlive-combined-basic-2021.20210408/bin/lualatex'
> ```

I could reproduce this problem on textext 1.4.0, but then I updated this PR to 1.8.1 and now I don't see it anymore \o/

jtojnar commented:
I tried rebasing onto nixos-unstable this but it no longer runs:

```
__main__.py: __main__.py: cannot execute binary file
```

possibly due to introduction of binary wrappers.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/how-to-define-an-inkscape-extension-with-python-dependencies/29384/3

raboof commented:
rebased and added a 'start script' to make sure the extension is invoked again.

No errors anymore, and the extension python code is indeed executed, but it seems to hang(?) without opening an edit window.

vale981 commented:
When running inkscape with textext from the nixpkgs of the pr, the textext window does not open for me and no error message is logged into the terminal :(

raboof commented:
> When running inkscape with textext from the nixpkgs of the pr, the textext window does not open for me and no error message is logged into the terminal :(

Yes, that confirms what I saw. I looked into it slightly but saw nothing obvious. Let's first update to 1.9.0, which requires #245431

vale981 commented:
Thanks for your efforts. I hacked together a solution with tkinter that kinda works but is super ugly and a pain in the neck as it builds a large number of python packages every time...

raboof commented:
okay, fixed two problems. This PR should be usable now (even without 1.3), if you can ignore a` <frozen importlib._bootstrap>:671: ImportWarning: DynamicImporter.exec_module() not found; falling back to load_module()` warning.

vale981 commented:
Can confirm: it works ) 
thank you so much

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/inkscape-extensions/33882/4

taha-yassine commented:
@jtojnar
Very sorry for the ping, but could you look into merging this as it seems ready? Thanks in advance!

vale981 commented:
@jtojnar I know that life can get busy some times. Could you maybe nominate somebody to have a look whether this can be merged?

roberth commented:
This has been ready for a while, I don't see any obvious problems and additions are low risk. I'm sure @raboof or any of the other contributors want to help out if anything turns out can still be improved.
